### PR TITLE
Swerve features

### PR DESCRIPTION
**Pull request by thetaspirit**
Basic odometry, brake/coast, and update to reset gyro

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -43,6 +43,13 @@
         public static final double MAX_SPEED_METERS_PER_SECOND = 3.627;
         public static final double MAX_MODULE_ANGULAR_SPEED_RADIANS_PER_SECOND = 8.76;
 
+        /**
+         * Multiply the rpm of the drive motor of a swerve wheel by this number to
+         * convert to meters per second.
+         */
+        public static final double DRIVE_WHEEL_RPM_TO_METERS_PER_SECOND = 1;
```
Comment (thetaspirit) [inline src/main/java/frc/robot/Constants.java]: This number needs to be calculated. I think `MAX_SPEED` is also outdated.

File: src/main/java/frc/robot/RobotContainer.java [hunk 2]
```diff
@@ -83,7 +84,14 @@ private void configureButtonBindings() {
...
-    new JoystickButton(m_driveController, Button.kStart.value).whenPressed(new ResetGyroCommand(m_swerveSubsystem));
+    // resets the gyro when the button is pressed
+    new JoystickButton(m_driveController, Button.kStart.value)
+        .whenPressed(new InstantCommand(m_swerveSubsystem::resetGyro, m_swerveSubsystem));
+    // Sets brake and coast mode
+    new JoystickButton(m_driveController, Button.kBumperLeft.value)
+        .whenPressed(() -> m_swerveSubsystem.setDriveIdleMode(IdleMode.kCoast));
+    new JoystickButton(m_driveController, Button.kBumperLeft.value)
+        .whenReleased(() -> m_swerveSubsystem.setDriveIdleMode(IdleMode.kBrake));
   }
```
Comment (TiaSinghania) [inline src/main/java/frc/robot/RobotContainer.java:94]: minor note, but we will have to properly clean up button bindings later. the left bumper is currently used for shooteroncommand in main.

File: src/main/java/frc/robot/commands/SwerveJoystickCommand.java [hunk 3]
```diff
@@ -29,19 +26,22 @@ public SwerveJoystickCommand(SwerveDriveSubsystem drivetrain, XboxController con
...
   public void initialize() {
-    
+
   }
 
   // Called every time the scheduler runs while the command is scheduled.
   @Override
   public void execute() {
     // Just think really hard about why these values are negated and flip-flopped.
     // Maybe use a whiteboard or piece of paper.
```
Comment (TiaSinghania) [inline src/main/java/frc/robot/commands/SwerveJoystickCommand.java]: should we write why they are negated and flipflopped in a javadoc so future people don't have to think really hard?

File: src/main/java/frc/robot/subsystems/SwerveDriveSubsystem.java [hunk 4]
```diff
@@ -137,14 +119,27 @@ public void move(double xSpeed, double ySpeed, double rotationSpeed, boolean isF
...
+    m_isTurning = rotationSpeed != 0;
   }
 
   /**
-   * Resets the gyro.
-   * Aka, sets the current heading of the robot to zero.
-   * TODO make sure this also updates odometry, <i>if needed</i>.
+   * Resets the gyro. Note that whatever command calls this method <b>will require
+   * the subsystem</b>. So no other commands can run on the subsystem while the
+   * gyro is being reset, including any drive control commands.
```
Comment (TiaSinghania) [inline src/main/java/frc/robot/subsystems/SwerveDriveSubsystem.java:174]: does this still apply for an instant command?

**Review by TiaSinghania — APPROVED** [review]

File: src/main/java/frc/robot/commands/SwerveJoystickCommand.java (hunk 3 above)
Comment (thetaspirit) [reply src/main/java/frc/robot/commands/SwerveJoystickCommand.java]: Lol that might be a good idea.  Or, in ur curriculum materials folder in the google drive, mention it?

File: src/main/java/frc/robot/subsystems/SwerveDriveSubsystem.java (hunk 4 above)
Comment (thetaspirit) [reply src/main/java/frc/robot/subsystems/SwerveDriveSubsystem.java:174]: Good point.  If you look at [the docs for that InstantCommand constructor](https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/first/wpilibj2/command/InstantCommand.html#%3Cinit%3E(java.lang.Runnable,edu.wpi.first.wpilibj2.command.Subsystem...)), the subsystem that I pass in as the second parameter is the subsystem that the command requires.

File: src/main/java/frc/robot/commands/SwerveJoystickCommand.java (hunk 3 above)
Comment (TiaSinghania) [reply src/main/java/frc/robot/commands/SwerveJoystickCommand.java]: yeah that could be part of the swerve explanation, but it shouldnt ONLY be there, this is javadoc material. doesn't need to be added now, but eventually we'll need to do a full cleanup

**Review by ProgrammingSR — CHANGES_REQUESTED** [review]
The new instance of the left bumper is redundant. Move .whenReleased() to the first instance. (Method Chaining) (Kanav and Lenny)

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 5]
```diff
@@ -40,8 +39,7 @@ public SwerveModule(CANSparkMax driveMotor, CANSparkMax turningMotor, double X,
     m_driveMotor = driveMotor;
     m_turningMotor = turningMotor;
     m_turningEncoder = turningEncoder;
-    m_turningPidController = new PIDController(Constants.SwerveConstants.MODULE_PID_P,
-        Constants.SwerveConstants.MODULE_PID_I, Constants.SwerveConstants.MODULE_PID_D);
+    m_turningPidController = new PIDController(0.3, 0, 0);
```
Comment (ProgrammingSR) [inline src/main/java/frc/robot/subsystems/SwerveModule.java:43]: Change the numbers back to constants using just something like SwerveConstants.MODULE_PID_P (Kanav and Lenny)

Comment (TiaSinghania) [reply src/main/java/frc/robot/subsystems/SwerveModule.java:43]: why do you want pid coefficients in constants? it makes tuning a pain, the coefficients are (almost) never used in the same place twice, and it doesn't correspond to any sort of hardware on the bot.

**Pull request merged**